File: langs/swift/Sources/VMx/Collections/TokenPagedComposition.swift
//
// TokenPagedComposition.swift — accumulated, forward-only token pagination.
//
import Foundation
import Combine

public final class TokenPagedComposition<TVM, TToken> {
    public typealias Page = ([TVM], TToken?)
    public typealias FetchNext = (TToken?) async throws -> Page
    public typealias PagesEqual = ([TVM], [TVM]) -> Bool

    private let fetchNext: FetchNext
    private let autoConstructOnAdd: Bool
    private let pagesEqual: PagesEqual
    private let stateQueue = DispatchQueue(label: "VMx.TokenPagedComposition.state")
    private var _items: [TVM] = []
    private var _currentToken: TToken?
    private var loadedOnce = false
    private var operationGeneration = 0
    private var disposed = false
    private let collectionChangedSubject = PassthroughSubject<CollectionChangedEvent, Never>()
    private let propertyChangedSubject = PassthroughSubject<String, Never>()
    private let commandChangedSubject = PassthroughSubject<Void, Never>()

    public lazy var loadMoreCommand: AsyncRelayCommand = AsyncRelayCommand.builder()
        .predicate { [weak self] in self?.hasMore == true && self?.isDisposed == false }
        .triggers(commandChangedSubject.eraseToAnyPublisher())
        .task { [weak self] in try await self?.loadMore() }
        .build()

    public lazy var refreshCommand: AsyncRelayCommand = AsyncRelayCommand.builder()
        .predicate { [weak self] in self?.isDisposed == false }
        .triggers(commandChangedSubject.eraseToAnyPublisher())
        .task { [weak self] in try await self?.refresh() }
        .build()

    public init(
        fetchNext: @escaping FetchNext,
        autoConstructOnAdd: Bool = false,
        pagesEqual: @escaping PagesEqual = { _, _ in false }
    ) {
        self.fetchNext = fetchNext
        self.autoConstructOnAdd = autoConstructOnAdd
        self.pagesEqual = pagesEqual
    }

    public var items: [TVM] { stateQueue.sync { _items } }

    public var currentToken: TToken? { stateQueue.sync { _currentToken } }

    public var hasMore: Bool {
        stateQueue.sync { !loadedOnce || _currentToken != nil }
    }

    private var isDisposed: Bool { stateQueue.sync { disposed } }

    public var collectionChanged: AnyPublisher<CollectionChangedEvent, Never> {
        collectionChangedSubject.eraseToAnyPublisher()
    }

    public var propertyChanged: AnyPublisher<String, Never> {
        propertyChangedSubject.eraseToAnyPublisher()
    }

    private func loadMore() async throws {
        let start = stateQueue.sync { () -> (disposed: Bool, token: TToken?, generation: Int) in
            guard !disposed else { return (true, nil, operationGeneration) }
            operationGeneration += 1
            return (false, _currentToken, operationGeneration)
        }
        guard !start.disposed else { return }

        let page = try await fetchNext(start.token)
        let isCurrent = stateQueue.sync {
            !disposed && start.generation == operationGeneration
        }
        guard isCurrent else { return }
        try constructIfNeeded(page.0)
        let committed = stateQueue.sync { () -> Bool in
            guard !disposed, start.generation == operationGeneration else { return false }
            _items.append(contentsOf: page.0)
            _currentToken = page.1
            loadedOnce = true
            return true
        }
        guard committed else { return }
        notifyResetIfLive()
    }

    private func refresh() async throws {
        let generation = stateQueue.sync { () -> Int? in
            guard !disposed else { return nil }
            operationGeneration += 1
            return operationGeneration
        }
        guard let generation else { return }
        let page = try await fetchNext(nil)
        let head = stateQueue.sync { () -> [TVM]? in
            guard !disposed, generation == operationGeneration else { return nil }
            return Array(_items.prefix(page.0.count))
        }
        guard let head else { return }
        let pagesMatch = pagesEqual(page.0, head)
        if !pagesMatch {
            try constructIfNeeded(page.0)
        }
        let committed = stateQueue.sync { () -> Bool in
            guard !disposed, generation == operationGeneration else { return false }
            if !pagesMatch {
                _items = page.0
            }
            _currentToken = page.1
            loadedOnce = true
            return true
        }
        guard committed else { return }
        if pagesMatch {
            notifyPropertiesIfLive()
        } else {
            notifyResetIfLive()
        }
    }

    private func constructIfNeeded(_ items: [TVM]) throws {
        guard autoConstructOnAdd else { return }
        for item in items {
            if let vm = item as? ComponentVMBase, !vm.isConstructed {
                try vm.construct()
            }
        }
    }

    private func notifyResetIfLive() {
        guard !isDisposed else { return }
        collectionChangedSubject.send(.reset())
        notifyPropertiesIfLive()
    }

    private func notifyPropertiesIfLive() {
        for name in ["items", "currentToken", "hasMore"] {
            guard !isDisposed else { return }
            propertyChangedSubject.send(name)
        }
        guard !isDisposed else { return }
        commandChangedSubject.send(())
    }

    public func dispose() {
        let shouldDispose = stateQueue.sync { () -> Bool in
            guard !disposed else { return false }
            disposed = true
            return true
        }
        guard shouldDispose else { return }
        loadMoreCommand.dispose()
        refreshCommand.dispose()
        collectionChangedSubject.send(completion: .finished)
        propertyChangedSubject.send(completion: .finished)
        commandChangedSubject.send(completion: .finished)
    }
}

File: langs/typescript/src/collections/tokenPagedComposition.ts
/**
 * TokenPagedComposition<TVM, TToken> — accumulated, forward-only token pagination.
 */
import { Subject } from "rxjs";
import type { Observable } from "rxjs";
import { AsyncRelayCommand } from "../commands/asyncRelayCommand.js";
import { ComponentVMBase } from "../components/componentVMBase.js";
import {
  makeCollectionChangedEvent,
  type CollectionChangedEvent,
} from "./collectionChangedEvent.js";

export interface TokenPage<TVM, TToken> {
  readonly items: readonly TVM[];
  readonly nextToken: TToken | null;
}

export interface TokenPagedCompositionOptions<TVM> {
  readonly autoConstructOnAdd?: boolean;
  readonly pagesEqual?: (left: readonly TVM[], right: readonly TVM[]) => boolean;
}

export class TokenPagedComposition<TVM, TToken> {
  readonly #fetchNext: (token: TToken | null) => Promise<TokenPage<TVM, TToken>>;
  readonly #autoConstructOnAdd: boolean;
  readonly #pagesEqual: (left: readonly TVM[], right: readonly TVM[]) => boolean;
  readonly #collectionChanged = new Subject<CollectionChangedEvent>();
  readonly #propertyChanged = new Subject<string>();
  readonly #commandChanged = new Subject<void>();
  readonly #loadMoreCommand: AsyncRelayCommand;
  readonly #refreshCommand: AsyncRelayCommand;
  #items: TVM[] = [];
  #currentToken: TToken | null = null;
  #loadedOnce = false;
  #operationGeneration = 0;
  #disposed = false;

  constructor(
    fetchNext: (token: TToken | null) => Promise<TokenPage<TVM, TToken>>,
    options: TokenPagedCompositionOptions<TVM> = {},
  ) {
    this.#fetchNext = fetchNext;
    this.#autoConstructOnAdd = options.autoConstructOnAdd ?? false;
    this.#pagesEqual = options.pagesEqual ?? shallowEqual;
    this.#loadMoreCommand = AsyncRelayCommand.builder()
      .predicate(() => this.hasMore && !this.#disposed)
      .triggers(this.#commandChanged)
      .task(async () => { await this.#loadMore(); })
      .build();
    this.#refreshCommand = AsyncRelayCommand.builder()
      .predicate(() => !this.#disposed)
      .triggers(this.#commandChanged)
      .task(async () => { await this.#refresh(); })
      .build();
  }

  get items(): TVM[] {
    return [...this.#items];
  }

  get currentToken(): TToken | null {
    return this.#currentToken;
  }

  get hasMore(): boolean {
    return !this.#loadedOnce || this.#currentToken !== null;
  }

  get loadMoreCommand(): AsyncRelayCommand {
    return this.#loadMoreCommand;
  }

  get refreshCommand(): AsyncRelayCommand {
    return this.#refreshCommand;
  }

  get collectionChanged(): Observable<CollectionChangedEvent> {
    return this.#collectionChanged.asObservable();
  }

  get propertyChanged(): Observable<string> {
    return this.#propertyChanged.asObservable();
  }

  async #loadMore(): Promise<void> {
    if (this.#disposed) return;
    const generation = ++this.#operationGeneration;
    const page = await this.#fetchNext(this.#currentToken);
    if (!this.#isOperationCurrent(generation)) return;
    this.#constructIfNeeded(page.items);
    if (!this.#isOperationCurrent(generation)) return;
    this.#items.push(...page.items);
    this.#currentToken = page.nextToken;
    this.#loadedOnce = true;
    this.#notifyReset();
  }

  async #refresh(): Promise<void> {
    if (this.#disposed) return;
    const generation = ++this.#operationGeneration;
    const page = await this.#fetchNext(null);
    if (!this.#isOperationCurrent(generation)) return;
    const head = this.#items.slice(0, page.items.length);
    const pagesMatch = this.#pagesEqual(page.items, head);
    if (!pagesMatch) this.#constructIfNeeded(page.items);
    if (!this.#isOperationCurrent(generation)) return;
    if (!pagesMatch) this.#items = [...page.items];
    this.#currentToken = page.nextToken;
    this.#loadedOnce = true;
    if (pagesMatch) this.#notifyProperties();
    else this.#notifyReset();
  }

  #isOperationCurrent(generation: number): boolean {
    return !this.#disposed && generation === this.#operationGeneration;
  }

  #constructIfNeeded(items: readonly TVM[]): void {
    if (!this.#autoConstructOnAdd) return;
    for (const item of items) {
      if (item instanceof ComponentVMBase && !item.isConstructed) {
        item.construct();
      }
    }
  }

  #notifyReset(): void {
    if (this.#disposed) return;
    this.#collectionChanged.next(makeCollectionChangedEvent("reset"));
    this.#notifyProperties();
  }

  #notifyProperties(): void {
    for (const name of ["items", "currentToken", "hasMore"]) {
      if (this.#disposed) return;
      this.#propertyChanged.next(name);
    }
    if (this.#disposed) return;
    this.#commandChanged.next();
  }

  dispose(): void {
    if (this.#disposed) return;
    this.#disposed = true;
    this.#loadMoreCommand.dispose();
    this.#refreshCommand.dispose();
    this.#collectionChanged.complete();
    this.#propertyChanged.complete();
    this.#commandChanged.complete();
  }
}

function shallowEqual<T>(left: readonly T[], right: readonly T[]): boolean {
  return left.length === right.length && left.every((item, index) => Object.is(item, right[index]));
}
